src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php

```diff
@@ -152,14 +142,22 @@ public function validate($entity, Constraint $constraint)
             reset($result);
         }
 
-        /* If no entity matched the query criteria or a single entity matched,
-         * which is the same as the entity being validated, the criteria is
-         * unique.
-         */
-        if (0 === count($result) || (1 === count($result) && $entity === ($result instanceof \Iterator ? $result->current() : current($result)))) {
+        if (0 === count($result)) {
             return;
         }
 
+        if (1 === count($result)) {
+            $match = $result instanceof \Iterator ? $result->current() : current($result);
+
+            if ($entity === $match) {
+                return;
+            }
+
+            if ($entity instanceof EntityDto && $entity->isForEntity($match)) {
```

Comment:
Probably as annotation option should be "better", wdyt? something like `isEqualToMethod`?

Reply:
Hmm maybe, this way would be more callback-constraint-like

Reply:
This approach will solve https://github.com/symfony/symfony/issues/22592#issuecomment-298375682